igvm/hypervisor_preferences.py

```diff
@@ -1,63 +1,93 @@
 """igvm - Hypervisor Selection Preferences
 
-This module contains preferences to select hypervisors.  Preferences return
-a value of any comparable datatype.  Only the return values of the same
-preference is compared with each other.  Smaller values mark hypervisors
-as more preferred.  Keep in mind that for booleans false is less than true.
-See sorted_hypervisors() function below for the details of sorting.
-
-Copyright (c) 2018 InnoGames GmbH
+This module contains preference classes to select Hypervisors. Preferences
+return either a score of range 0.0 - 1.0 inclusive, or a boolean value, which
+is only there for convenience and converts to float.
+A higher score means that a Hypervisor should be favored, while a low one
+discourages a HV. If False or 0.0 is given, that means a HV does not fulfill
+the requirements at all and is excluded.
+
+Copyright (c) 2020 InnoGames GmbH
 """
-# This module contains the preferences as simple classes.  We try to keep
-# them reusable, even though most of them are not reused.  Some of the classes
-# are so simple that they could as well just be a function, but kept
-# as classes to have a consistent style.
-
 from logging import getLogger
-
-from igvm.utils import LazyCompare
+from typing import Union, List
 
 log = getLogger(__name__)
 
 
-class InsufficientResource(object):
-    """Check a resource of hypervisor would be sufficient"""
-    def __init__(self, hv_attribute, vm_attribute, multiplier=1, reserved=0):
-        self.hv_attribute = hv_attribute
-        self.vm_attribute = vm_attribute
-        self.multiplier = multiplier  # TODO Use identical units in Serveradmin
-        self.reserved = reserved
-
-    def __repr__(self):
+class HypervisorPreference:
+    """The base class for all HV preferences."""
+
+    def get_score(self, vm, hv) -> Union[float, bool]:
+        """Calculates a preference value to indicate how good a HV fits.
+
+        @param vm: The VM object to check against the HV.
+        @param hv: The HV object to check against the VM.
+        @return: A float or a bool. The bool is just for convenience and will
+                 be converted to a float. A value of 0.0 means the HV does not
+                 fulfill the requirement at all. But a value of 1.0 means it is
+                 a very good candidate for this particular preference.
+        @rtype: Union[float, bool]
+        """
+        raise NotImplementedError('get_preference is not implemented')
```

Comment:
I think this log message should be `get_score` to reflect the function name

Reply:
good catch

Reply:
i made the class `ABC` instead